src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,82 +34,244 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
     }
 
     private void processInput(String userInput) {
+        filterMilestonesByQuerySentence(allMilestones, userInput);
+        populateBestMatchingMilestones(userInput);
         if (userInput.isEmpty()) return;
+        toggleFirstMilestone(bestMatchingMilestones);
+    }
 
-        String[] userInputWords = userInput.split(" ");
+    private void filterMilestonesByQuerySentence(List<PickerMilestone> milestones, String querySentence) {
+        String[] userInputWords = querySentence.split(" ");
         Stream.of(userInputWords)
-                .forEach(this::filterMilestones);
+                .forEach(queryWord -> filterMilestonesByQueryWord(milestones, queryWord));
+    }
 
-        toggleFirstMatchingMilestone(currentMilestonesList);
+    private void filterMilestonesByQueryWord(List<PickerMilestone> milestones, String queryWord) {
+        milestones.stream()
+                .forEach(milestone -> {
+                    if (milestone.isMatching()) milestone.setMatching(isMatchingQuery(milestone, queryWord));
+                });
     }
 
     /**
-     * Toggles the milestone that has milestoneName as its exact name
-     * <p>
-     * This method is case-sensitive
+     * Toggles the selection status of the milestone that has milestoneName as its exact name
```

Comment:
Let's reword this. `Toggles the selection of the milestone which name is exactly milestoneName. This method is case-sensitive` (don't give line break as it may be skipped).